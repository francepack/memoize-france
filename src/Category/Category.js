import React, { Component } from 'react';
import Question from '../Question/Question';
import Feedback from '../Feedback/Feedback';

export default class Category extends Component {
  constructor(props) {
    super(props);
    this.state = {
      categoryQuestions: [],
      selectedQuestion: {},
      questionCount: 0,
      incorrectCount: 0,
      correctCount: 0,
      currentQuestionCorrect: false,
      displayQuestion: false,
      showFeedback: false
    }
  }

  showQuestion = () => {
    this.setState({
      selectedQuestion: this.state.categoryQuestions[this.state.questionCount]
    })
    this.toggleShowQuestion()
  }

  toggleShowQuestion = () => {
    this.setState({ displayQuestion: !this.state.displayQuestion })
  }

  toggleShowFeedback = () => {
    this.setState({ showFeedback: !this.state.showFeedback })
  }

  evaluateQuestion = (e) => {
    if (e.target.innerText === this.state.selectedQuestion.answer) {
      this.setState({ 
        correctCount: this.state.correctCount + 1,
        questionCount: this.state.questionCount + 1,
        currentQuestionCorrect: true 
      });
    } else {
      this.setState({
        incorrectCount: this.state.incorrectCount + 1,
        questionCount: this.state.questionCount + 1,
        currentQuestionCorrect: false
      });
      this.props.collectMissedQuestions(this.state.selectedQuestion.id);
    }
    this.toggleShowQuestion();
    this.toggleShowFeedback();
  }

  componentDidMount() {
    if (this.props.category) {
      let categoryQuestions = this.props.questions.filter(question => {
        return question.category === this.props.category
      })
    this.setState({ categoryQuestions: categoryQuestions })
    } else {
      this.setState({ categoryQuestions: this.props.storage})
    }
  }

  render() {
    return(
      <div className="category-box">
      { this.state.questionCount !== this.state.categoryQuestions.length &&
        <article>
          { this.props.category &&
          <h2> {this.props.category} </h2>
          }
          { !this.props.category &&
          <div>
            <h2> Review </h2>
            <p className="review">Revisit the questions that you previously have missed!</p>
          </div>
          }
          <section className="category-stats">
            <p className="category-completion">You've completed {this.state.questionCount}/{this.state.categoryQuestions.length} questions in this category.</p>
            <p className="category-misses">You've missed {this.state.incorrectCount} questions.</p>
          </section>
          <button onClick={this.showQuestion} className="question-btn">Fire Question!</button>
          <Question 
            displayQuestion={this.state.displayQuestion}
            selectedQuestion={this.state.selectedQuestion}
            evaluateQuestion={this.evaluateQuestion}
          />
          <Feedback 
            showFeedback={this.state.showFeedback}
            questionInfo={this.state.selectedQuestion.link}
            toggleFeedback={this.toggleShowFeedback}
            questionFeedback={this.state.currentQuestionCorrect}
          />  
        </article>
      }
      { this.state.questionCount === this.state.categoryQuestions.length &&
      <div className="finish-category-summary">
       { !this.props.category &&
          <div>
            <h2> Review </h2>
            <p className="review">Revisit the questions that you previously have missed!</p>
          </div>
          }
        <p>All questions attempted.</p>
        <span>You answered {this.state.correctCount} of {this.state.categoryQuestions.length} correctly.</span>
      </div>
      }
      </div>
    )
  }
}