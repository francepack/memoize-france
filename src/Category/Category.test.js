import React from 'react';
import Category from './Category';
import { shallow } from 'enzyme';

import {configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

//made all the categories the same vale for these tests
const mockData= [ {id:'1', category:'People', question:'Who is it?', options:['me', 'you', 'neither', 'both'], answer:'you', link:'http://plcaeholder.com'}, {id:'2', category:'People', question:'How old is my nephew?', options:['1', '2', '3', '4'], answer:'3', link:'http://plcaeholder.com'}, {id:'3', category:'People', question:'What is the best place to eat in Denver?', options:['Casa Bonita', 'Chipotle', 'Denver Teds', 'Birdcall'], answer:'Birdcall', link:'http://plcaeholder.com'} ];
const mockCategories= [ 'People', 'Age', 'Denver' ];
const showQuestion = jest.fn();



describe('Category', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Category
        category={mockCategories}
        questions={mockData}
      />
    )
  });

  it('should match snapshot', () => {
    wrapper.setState({'categoryQuestions': mockData});
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot', () => {
    wrapper.setState({'categoryQuestions': mockData, 'questionCount': 3});
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default states', () => {
    expect(wrapper.state()).toEqual({
      categoryQuestions: [], 
      selectedQuestion: {}, 
      questionCount: 0,
      incorrectCount: 0,
      correctCount: 0,
      currentQuestionCorrect: false,
      displayQuestion: false,
      showFeedback: false
    });
  });

  it('should update selectedQuestion state when showQuestion is invoked', () => {
    wrapper.setState({'categoryQuestions': mockData});
    expect(wrapper.state('selectedQuestion')).toEqual({});
    wrapper.instance().showQuestion();
    expect(wrapper.state('selectedQuestion')).toEqual({id:'1', category:'People', question:'Who is it?', options:['me', 'you', 'neither', 'both'], answer:'you', link:'http://plcaeholder.com'});
  });

  it('should invoke showQuestion when the Fire Question button is clicked', () => {
    wrapper.setState({'categoryQuestions': mockData, 'questionCount': 0});
    wrapper.find('.question-btn').simulate('click', { preventDefault: () => {}});
    expect(showQuestion).toBeCalled(); 
  });

  it('should update displayQuestion state when toggleShowQuestion is invoked', () => {
    expect(wrapper.state('displayQuestion')).toEqual(false);
    wrapper.instance().toggleShowQuestion();
    expect(wrapper.state('displayQuestion')).toEqual(true);
  });

  it('should update showFeedback state when toggleShowFeedback is invoked', () => {
    expect(wrapper.state('showFeedback')).toEqual(false);
    wrapper.instance().toggleShowFeedback();
    expect(wrapper.state('showFeedback')).toEqual(true);
  });

  // it('shold update questionCount and correct/incorrectCount states when evaluateQuestion is invoked', () => {
  //   expect(wrapper.state('questionCount')).toEqual(0);
  //   expect(wrapper.state('correctCount')).toEqual(0);
  //   expect(wrapper.state('incorrectCount')).toEqual(0);
  //   expect(wrapper.state('currentQuestionCorrect')).toEqual(false);
  //   wrapper.setState({'selectedQuestion': {id:'1', category:'People', question:'Who is it?', options:['me', 'you', 'neither', 'both'], answer:'you', link:'http://plcaeholder.com'});
  //   evaluateQuestion();
  //   expect(wrapper.state('questionCount')).toEqual(1);
  //   expect(wrapper.state('correctCount')).toEqual(1);
  //   expect(wrapper.state('incorrectCount')).toEqual(0);
  //   expect(wrapper.state('currentQuestionCorrect')).toEqual(true);

  // });

});  

