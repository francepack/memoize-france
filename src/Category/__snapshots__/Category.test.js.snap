// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category should match snapshot 1`] = `
<div
  className="category-box"
>
  <article>
    <h2>
       
      People
      Age
      Denver
       
    </h2>
    <section
      className="category-stats"
    >
      <p
        className="category-completion"
      >
        You've completed 
        0
        /
        3
         questions in this category.
      </p>
      <p
        className="category-misses"
      >
        You've missed 
        0
         questions.
      </p>
    </section>
    <button
      className="question-btn"
      onClick={[Function]}
    >
      Fire Question!
    </button>
    <Question
      displayQuestion={false}
      evaluateQuestion={[Function]}
      selectedQuestion={Object {}}
    />
    <Feedback
      questionFeedback={false}
      showFeedback={false}
      toggleFeedback={[Function]}
    />
  </article>
</div>
`;

exports[`Category should match snapshot 2`] = `
<div
  className="category-box"
>
  <div
    className="finish-category-summary"
  >
    <p>
      All questions attempted.
    </p>
    <p>
      You answered 
      0
       of 
      3
       correctly.
    </p>
  </div>
</div>
`;
